Add flags for CPU/Memory requests/limits

JJJJJones opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->

Config Reloader's CPU requests/CPU limits and Memory requests/Memory limits are set whatever is given to their respective flags, this allows them to be independently set (EG CPU requests of 100m but no CPU limits) #3825

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:feature

Unless you choose release-note:none, please add a release note.
-->
**Release Note Template (will be copied)**

```release-note:REPLACEME
[FEATURE] Allow Config Reloader CPU requests/limits and memory requests/limits to be set independently #3825 
```

JJJJJones commented:
This was made with the assistance of @rmartinez3.

lilic commented on cmd/operator/main.go:
```diff
@@ -171,7 +171,11 @@ func init() {
+	flagset.StringVar(&cfg.ReloaderConfig.CPURequest, "config-reloader-cpu-req", "", "Config Reloader CPU request. Value \"0\" disables it and causes no request to be configured.")
+	flagset.StringVar(&cfg.ReloaderConfig.CPULimit, "confiig-reloader-cpu-limit", "", "Config Reloader CPU limit. Value \"0\" disables it and causes no limit to be configured.")
 	flagset.StringVar(&cfg.ReloaderConfig.Memory, "config-reloader-memory", "50Mi", "Config Reloader Memory request & limit. Value \"0\" disables it and causes no request/limit to be configured.")
+	flagset.StringVar(&cfg.ReloaderConfig.MemoryRequest, "config-reloader-memory-req", "", "Config Reloader Memory request. Value \"0\" disables it and causes no request to be configured.")
```
Lets say you use both, we should probably at least write that config-reloader-memory supersedes this flag, correct?

JJJJJones replied:
I forgot to leave a comment that I was unsure what the desired comment would be. Perhaps "Setting this flag invalidates flag `<relevant_flag>`"?

lilic commented on pkg/operator/statefulset.go:
```diff
@@ -75,13 +75,38 @@ func CreateConfigReloader(
-		resources.Limits[v1.ResourceCPU] = resource.MustParse(config.CPU)
-		resources.Requests[v1.ResourceCPU] = resource.MustParse(config.CPU)
+
+	// Prioritize specific flags over generic ones
```
Lets add some tests for these use cases, unit tests is fine maybe we can separate it out to its standalone function, thanks!

lilic commented on pkg/operator/statefulset.go:
```diff
@@ -75,13 +75,38 @@ func CreateConfigReloader(
+		}
+	}
+
+	if config.CPURequest != "0" {
```
Should this not go under the `if config.CPU != ""` case? same for memory below?

JJJJJones replied:
Currently, no memory/CPU limits are achieved in the code by not specifying a value for resources.Limits[v1.ResourceMemory] and resources.Limits[v1.ResourceCPU] respectively (as triggered by config.CPU and config.Memory being "0"). So this follows that by not specifying a value for the specific CPU/Memory request/limit if the value in the relevant config is "0".

lilic reviewed (commented):
Thanks for the PR!

@simonpasquier please have a look and what you think of this proposal, thanks!

simonpasquier commented on pkg/operator/config.go:
```diff
@@ -45,9 +45,13 @@ type Config struct {
-	CPU    string
-	Memory string
-	Image  string
+	CPU           string
```
I would get rid of the `CPU` field (same for `Memory`) and move the logic of choosing between the global value and request/limit values in main().

JJJJJones replied:
I'll make a temporary variable to replace the now removed `cfg.ConfigReloader.CPU` and `cfg.ConfigRelaoder.Memory`.

JJJJJones replied:
A number of tests break with the removal of ReloaderConfig.CPU and ReloaderConfig.Memory. Correcting...

JJJJJones replied:
Fixed.

simonpasquier reviewed (commented):
IMO we should support only one way for setting limits/requests eventually so I would recommend to add a deprecation note on the `--config-reloader-cpu` and `--config-reloader-memory` flags.

cc @prometheus-operator/prometheus-operator-reviewers for more input.

JJJJJones commented:
> IMO we should support only one way for setting limits/requests eventually so I would recommend to add a deprecation note on the `--config-reloader-cpu` and `--config-reloader-memory` flags.

@simonpasquier I'm not opposed to that at all and I think that's not the worst possible idea just to avoid potential confusion (as an added benefit it simplifies the code in `pkg/operator/statefulset.go`) but I wasn't sure what was desired and didn't want to start breaking things with behavioral changes so I left those flags in and just wrote code to handle usage of both.

JJJJJones commented on cmd/operator/main.go:
```diff
@@ -171,7 +171,11 @@ func init() {
 	// specified.
 	flagset.StringVar(&cfg.ReloaderConfig.Image, "prometheus-config-reloader", operator.DefaultPrometheusConfigReloaderImage, "Prometheus config reloader image")
 	flagset.StringVar(&cfg.ReloaderConfig.CPU, "config-reloader-cpu", "100m", "Config Reloader CPU request & limit. Value \"0\" disables it and causes no request/limit to be configured.")
+	flagset.StringVar(&cfg.ReloaderConfig.CPURequest, "config-reloader-cpu-req", "", "Config Reloader CPU request. Value \"0\" disables it and causes no request to be configured.")
```
We would also want to change the memory request flag to be "config-reloader-memory-request".

I chose to go with req instead of request just to keep the flag length down. Dunno how much concern there is around that. If none, then i'll change the flags to "config-reloader-cpu-request" and "config-reloader-memory-request".

lilic replied:
There is no concern, request is better for both agreed.

JJJJJones replied:
Sounds good, changed both

simonpasquier commented:
@JJJJJones sure no problem, that's the whole purpose of having code reviews :)
We usually support both flags a couple of releases so users can update their manifests and then we remove the deprecated flags (see #3773 and #3457 for instance).

Lets wait for other opinions though.

lilic commented:
Agreed with Simon, I would vote for deprecation as well, but maybe support for at least two releases. @simonpasquier do you remember what we did with previous flags?

simonpasquier commented:
yes we said that the deprecated flags would be removed after 2 releases (e.g. deprecation announced in 0.43 and removed in 0.45)

lilic commented:
Sounds good to me, @JJJJJones do you mind applying the suggestion of marking current flags as deprecated in the flag help text. Unless @prometheus-operator/kube-prometheus-reviewers have any objections.

JJJJJones commented:
Want me to go ahead and draw up the PR for when those flags flags are removed as well?

lilic commented:
Opening an issue after we merge this PR and assigning to youself sounds perfect as if we make some changes rebasing might be harder.

s-urbaniak commented:
Agreed regarding deprecation notice.

JJJJJones commented:
@lilic @s-urbaniak @simonpasquier 

I've added the deprecation notices, updated the flags to request, removed CPU and Memory from `operator.ReloaderConfig`, and moved some of the logic out of `pkg/operator/statefulset.go`. The existing tests have been updated for correctness with the removal of CPU and Memory from `operator.ReloaderConfig`.

lilic commented:
Nice work and progress, just ping when its ready for final review, thank you! 🎉

JJJJJones commented:
@lilic the tests have been written up. Let me know if you need anything else.

JJJJJones commented on pkg/alertmanager/statefulset_test.go:
```diff
@@ -33,9 +33,11 @@ import (
+			CPULimit:      "100m",
+			MemoryRequest: "50Mi",
+			MemoryLimit:   "50Mi",
 		},
```
All tests that specify resources utilize "50Mi" for memory so I've bumped the defaultTestConfig to match.

JJJJJones commented on pkg/thanos/statefulset_test.go:
```diff
@@ -64,9 +67,7 @@ func TestStatefulSetLabelingAndAnnotations(t *testing.T) {
-		Spec: monitoringv1.ThanosRulerSpec{
-			QueryEndpoints: emptyQueryEndpoints,
-		},
+		Spec: monitoringv1.ThanosRulerSpec{QueryEndpoints: emptyQueryEndpoints},
```
The test files are already aggressively long, may as well stay consistent and opt for one liners where possible.

JJJJJones commented on pkg/prometheus/statefulset_test.go:
```diff
@@ -1160,53 +1273,187 @@ func TestSidecarsNoCPULimits(t *testing.T) {
 		ReloaderConfig: operator.ReloaderConfig{
-			CPU:    "0",
-			Memory: "50Mi",
-			Image:  "quay.io/coreos/prometheus-config-reloader:latest",
```
This was the only place with the usage of prometheus-config-reloader from coreos and not prometheus-operator. I assume this was a mistake?

simonpasquier replied:
good catch!

JJJJJones commented:
@lilic I realize my last message was a little unclear. I'd say the MR is ready for final reviews.

simonpasquier commented on cmd/operator/main.go:
```diff
@@ -189,6 +197,26 @@ func init() {
+		}
+		if cfg.ReloaderConfig.CPULimit == "" {
+			cfg.ReloaderConfig.CPULimit = rcCPU
+		}
```
I would log a deprecation warning if rcCPU is different than `100m` so folks using the flag are potentially aware of the deprecation.

JJJJJones replied:
I can't address those then and there with logger so I baked the warning check into Main(). Let me know if there's a preferred alternative solution.

simonpasquier commented on cmd/operator/main.go:
```diff
@@ -189,6 +197,26 @@ func init() {
+		}
+		if cfg.ReloaderConfig.MemoryLimit == "" {
+			cfg.ReloaderConfig.MemoryLimit = rcMemory
+		}
```
same here

simonpasquier commented on pkg/operator/statefulset.go:
```diff
@@ -75,13 +75,18 @@ func CreateConfigReloader(
-		resources.Requests[v1.ResourceCPU] = resource.MustParse(config.CPU)
+
+	if config.CPURequest != "0" {
+		resources.Requests[v1.ResourceCPU] = resource.MustParse(config.CPURequest)
```
Side note for a following PR: we should validate the values earlier in the process rather than panic'ing so late :)

JJJJJones replied:
I'm not sure what you mean by panicking. There is no zero/nil/null value for resources. The only way to for a container to have no CPU/memory requests/limits is to not set them. These checks are how we accomplish that.

simonpasquier replied:
I mean that [`resource.MustParse()`](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/resource#MustParse) will panic if the provided value isn't a valid quantity.

JJJJJones replied:
That makes sense. Want me to add to init() some code to do the check?

It would look like this

```go
	// ...
	flagset.StringVar(&cfg.PromSelector, "prometheus-instance-selector", "", "Label selector to filter Prometheus Custom Resources to watch.")
	flagset.StringVar(&cfg.AlertManagerSelector, "alertmanager-instance-selector", "", "Label selector to filter AlertManager Custom Resources to watch.")
	flagset.StringVar(&cfg.ThanosRulerSelector, "thanos-ruler-instance-selector", "", "Label selector to filter ThanosRuler Custom Resources to watch.")
	flagset.StringVar(&cfg.SecretListWatchSelector, "secret-field-selector", "", "Field selector to filter Secrets to watch")

	// Prioritize specific flags over generic ones
	// If config.CPU has been set but not config.CPURequest/config.CPULimit, default "" to config.CPU
	// If config.Memory has been set but not config.MemoryRequest or config.MemoryLimit, default "" to config.Memory
	if rcCPU != "" {
		if cfg.ReloaderConfig.CPURequest == "" {
			cfg.ReloaderConfig.CPURequest = rcCPU
		}
		if cfg.ReloaderConfig.CPULimit == "" {
			cfg.ReloaderConfig.CPULimit = rcCPU
		}
	}
	if rcMemory != "" {
		if cfg.ReloaderConfig.MemoryRequest == "" {
			cfg.ReloaderConfig.MemoryRequest = rcMemory
		}
		if cfg.ReloaderConfig.MemoryLimit == "" {
			cfg.ReloaderConfig.MemoryLimit = rcMemory
		}
	}
	resource.MustParse(cfg.ReloaderConfig.CPULimit)
	resource.MustParse(cfg.ReloaderConfig.CPURequest)
	resource.MustParse(cfg.ReloaderConfig.MemoryLimit)
	resource.MustParse(cfg.ReloaderConfig.MemoryRequest)
}
```

simonpasquier replied:
It would be great! Instead of MustParse(), you could call ParseQuantity(), log the error and return from Main().

JJJJJones replied:
I'll get the change committed.

JJJJJones replied:
Change has been committed.

simonpasquier reviewed (commented):
lgtm, just one ask to add a deprecation warning.

simonpasquier reviewed (approved):
lgtm, thanks! I'll let @lilic and @s-urbaniak  a chance to chime in before we merge it.

s-urbaniak commented:
lgtm as well, thank you! leaving the final pair of eyes to @lilic

lilic commented on cmd/operator/main.go:
```diff
@@ -170,8 +177,14 @@ func init() {
+	flagset.StringVar(&rcCPU, "config-reloader-cpu", defaultReloaderCPU, "Config Reloader CPU request & limit. Value \"0\" disables it and causes no request/limit to be configured. Deprecated, it will be removed in v0.48.0.")
+	flagset.StringVar(&cfg.ReloaderConfig.CPURequest, "config-reloader-cpu-request", "", "Config Reloader CPU request. Value \"0\" disables it and causes no request to be configured. Flag overrides `--config-reloader-cpu` value for the CPU request")
+	flagset.StringVar(&cfg.ReloaderConfig.CPULimit, "config-reloader-cpu-limit", "", "Config Reloader CPU limit. Value \"0\" disables it and causes no limit to be configured. Flag overrides `--config-reloader-cpu` for the CPU limit")
+	// TODO(JJJJJones): remove the '--config-reloader-memory' flag before releasing v0.48.
```
Same question as above :)

lilic reviewed (approved):
Mainly one question that can be done as a follow up! Thank you!!

JJJJJones commented on cmd/operator/main.go:
```diff
@@ -170,8 +177,14 @@ func init() {
 	flagset.StringVar(&cfg.ReloaderConfig.Image, "prometheus-config-reloader", operator.DefaultPrometheusConfigReloaderImage, "Prometheus config reloader image")
-	flagset.StringVar(&cfg.ReloaderConfig.CPU, "config-reloader-cpu", "100m", "Config Reloader CPU request & limit. Value \"0\" disables it and causes no request/limit to be configured.")
-	flagset.StringVar(&cfg.ReloaderConfig.Memory, "config-reloader-memory", "50Mi", "Config Reloader Memory request & limit. Value \"0\" disables it and causes no request/limit to be configured.")
+	// TODO(JJJJJones): remove the '--config-reloader-cpu' flag before releasing v0.48.
```
I'll make one.

JJJJJones replied:
#3883
